packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -85,123 +109,507 @@ const propTypes = {
+  const theme = React.useContext(ThemeContext);
+
+  const { onCalendarButtonClick, shouldShowPicker } = customProps;
+  delete customProps.onCalendarButtonClick;
+  delete customProps.shouldShowPicker;
+
+  const additionalInputProps = { ...customProps, ...inputAttributes };
+  const momentDateFormat = DateUtil.getFormatByLocale(intl.locale);
+  const dateValue = DateUtil.convertToISO8601(value, momentDateFormat);
+  const dateFormatOrder = intl.formatMessage({ id: 'Terra.datePicker.dateFormatOrder' });
+
+  const setDayRef = useCallback(node => {
+    dayInputRef.current = node;
+    if (inputRefCallback && dateFormatOrder === 'YMD') {
+      inputRefCallback(node);
+    }
+  }, [dateFormatOrder, inputRefCallback]);
+
+  const setMonthRef = useCallback(node => {
+    monthInputRef.current = node;
+  }, []);
+
+  const setYearRef = useCallback(node => {
+    yearInputRef.current = node;
+    if (inputRefCallback && (dateFormatOrder === 'MDY' || dateFormatOrder === 'DMY')) {
+      inputRefCallback(node);
+    }
+  }, [dateFormatOrder, inputRefCallback]);
+
+  useEffect(() => {
+    if (shouldShowPicker && onClick) {
+      onClick();
+    }
+  }, [shouldShowPicker, onClick]);
+
+  // Sets the date state based on the passed in value prop, or if it changes via a calendar click.
+  useEffect(() => {
+    if (DateUtil.isValidDate(value, momentDateFormat)) {
+      const parseDate = DateUtil.getDateInputValues('YMD', dateValue, '-');
+      dateDispatch(parseDate);
+    } else if (dateValue === '') {
+      dateDispatch({ day: '', month: '', year: '' });
+    }
+  }, [momentDateFormat, dateValue, value]);
+
+  /**
+   * Moves focus to the correct input depending on date ordering. Focus changing is
+   * disabled if a complete date has been entered in order to make single input
+   * corrections easier, and is re-enabled if the whole date is erased.
+   * @param {string} inputValue - The value from the current input.
+   * @param {number} type - The input type, based on DateUtil.inputType.
+   * @param {string} dateOrder - Date variant based on the currently supported date
+   *                      orders: 'MDY', 'DMY', or 'YMD'.
+   */
+  const moveFocusOnChange = (inputValue, type, dateOrder) => {
+    if (dateOrder === 'MDY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.MONTH) {
+          dayInputRef.current.focus();
+        } else {
+          yearInputRef.current.focus();
+        }
+      }
+    } else if (dateOrder === 'DMY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.DAY) {
+          monthInputRef.current.focus();
+        } else {
+          yearInputRef.current.focus();
+        }
+      }
+    } else if (dateOrder === 'YMD') {
+      if (inputValue.length === 4) {
+        monthInputRef.current.focus();
+      } else if (inputValue.length === 2 && type === DateUtil.inputType.MONTH) {
+        dayInputRef.current.focus();
+      }
+    }
+  };
```

Comment:
I am not sure whether it works for all scenarios. so can we rewrite the above logic like this ?

`if (dateOrder === 'MDY' || dateOrder === 'DMY' || dateOrder === 'YMD') {
     if (inputValue.length === 2) {
          if (type === DateUtil.inputType.MONTH && (dateOrder === 'MDY' ||  dateOrder === 'YMD' )) {
                dayInputRef.current.focus();
          } else if (type === DateUtil.inputType.DAY && dateOrder === 'DMY' ) {
                monthInputRef.current.focus();
          } else if (dateOrder === 'MDY' || dateOrder === 'DMY') {
                 yearInputRef.current.focus();
           }
     } else if (inputValue.length === 4). { 
          monthInputRef.current.focus();
     }
  }
`

Reply:
That can also be an alternative, but the reason for going for the current implementation was to avoid repetitive checks of dateOrder